state-sync/storage-service/server/src/lib.rs

```diff
@@ -356,6 +356,85 @@ impl StorageReader {
+            if latest_version > pruning_window {
+                // lowest_account_version = latest_version - pruning_window + 1;
+                let mut lowest_account_version =
+                    latest_version.checked_sub(pruning_window).ok_or_else(|| {
+                        Error::UnexpectedErrorEncountered(
+                            "Lowest account states version has overflown!".into(),
+                        )
+                    })?;
+                lowest_account_version =
+                    lowest_account_version.checked_add(1).ok_or_else(|| {
+                        Error::UnexpectedErrorEncountered(
+                            "Lowest account states version has overflown!".into(),
+                        )
+                    })?;
+
+                // Create the account range
+                let account_range = CompleteDataRange::new(lowest_account_version, latest_version)
+                    .map_err(|error| Error::UnexpectedErrorEncountered(error.to_string()))?;
+                Ok(Some(account_range))
+            } else {
+                // The pruner hasn't kicked in yet, so return the transactions range
+                Ok(*transactions_range)
+            }
+        } else {
+            // There is no pruning window, return the transactions range
+            Ok(*transactions_range)
+        }
+    }
+
+    /// Returns the transaction range held in the database (lowest to highest).
+    fn fetch_transaction_range(
+        &self,
+        latest_version: Version,
+    ) -> Result<Option<CompleteDataRange<Version>>, Error> {
+        let first_transaction_version = self
+            .storage
+            .get_first_txn_version()
+            .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
+        if let Some(first_transaction_version) = first_transaction_version {
+            let transaction_range =
+                CompleteDataRange::new(first_transaction_version, latest_version)
+                    .map_err(|error| Error::UnexpectedErrorEncountered(error.to_string()))?;
+            Ok(Some(transaction_range))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Returns the transaction output range held in the database (lowest to highest).
+    fn fetch_transaction_output_range(
+        &self,
+        latest_version: Version,
+    ) -> Result<Option<CompleteDataRange<Version>>, Error> {
+        let first_output_version = self
+            .storage
+            .get_first_write_set_version()
+            .map_err(|error| Error::StorageErrorEncountered(error.to_string()))?;
+        if let Some(first_output_version) = first_output_version {
+            let output_range = CompleteDataRange::new(first_output_version, latest_version)
+                .map_err(|error| Error::UnexpectedErrorEncountered(error.to_string()))?;
+            Ok(Some(output_range))
+        } else {
+            Ok(None)
+        }
+    }
```

Comment:
These are exactly the same except for the storage `get` command right?  Maybe can merge them.

Reply:
Gonna leave them as-is for now. But, I'll probably end up coming back to it when I make more modifications.